monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala

```diff
@@ -0,0 +1,218 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.operators
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.locks.ReentrantLock
+
+import monix.eval.Task
+import monix.execution.Ack.{Continue, Stop}
+import monix.execution.cancelables.{CompositeCancelable, SingleAssignCancelable}
+import monix.execution.misc.{AsyncSemaphore, NonFatal}
+import monix.execution.{Ack, Cancelable, CancelableFuture}
+import monix.reactive.observers.{BufferedSubscriber, Subscriber}
+import monix.reactive.{Observable, OverflowStrategy}
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+private[reactive] final class MapParallelOrderedObservable[A, B](
+  source: Observable[A],
+  parallelism: Int,
+  f: A => Task[B],
+  overflowStrategy: OverflowStrategy[B]) extends Observable[B] {
+
+  override def unsafeSubscribeFn(out: Subscriber[B]): Cancelable = {
+    if (parallelism <= 0) {
+      out.onError(new IllegalArgumentException("parallelism > 0"))
+      Cancelable.empty
+    }
+    else if (parallelism == 1) {
+      // optimization for one worker
+      new MapTaskObservable[A, B](source, f).unsafeSubscribeFn(out)
+    }
+    else {
+      val composite = CompositeCancelable()
+      val subscription = new MapAsyncParallelSubscription(out, composite)
+      composite += source.unsafeSubscribeFn(subscription)
+      subscription
+    }
+  }
+
+  private final class MapAsyncParallelSubscription(
+    out: Subscriber[B], composite: CompositeCancelable) extends Subscriber[A] with Cancelable { self =>
+
+    implicit val scheduler = out.scheduler
+    // Ensures we don't execute more than a maximum number of tasks in parallel
+    private[this] val semaphore = AsyncSemaphore(parallelism)
+    // Reusable instance for releasing permits on cancel, but
+    // it's debatable whether this is needed, since on cancel
+    // everything gets canceled at once
+    private[this] val releaseTask = Task.eval(semaphore.release())
+    // Buffer with the supplied  overflow strategy.
+    private[this] val buffer = BufferedSubscriber[B](out, overflowStrategy)
+
+    // Flag indicating whether a final event was called, after which
+    // nothing else can happen. It's a very light protection, as
+    // access to it is concurrent and not synchronized
+    private[this] var isDone = false
+    // Turns to `Stop` when a stop acknowledgement is observed
+    // coming from the `buffer` - this indicates that the downstream
+    // no longer wants any events, so we must cancel
+    private[this] var lastAck: Ack = Continue
+    // Buffer for signaling new elements downstream preserving original order
+    private[this] val queue = new ConcurrentLinkedQueue[CancelableFuture[B]]
+    // This lock makes sure that only one thread at the time sends processed elements downstream
+    private[this] val sendDownstreamLock = new ReentrantLock()
```

Comment:
What is the advantage of a `ReentrantLock`, versus intrinsic locks with `synchronize`?

The later has the advantage that the JVM can do escape analysis in some instances and eliminate it. I once had a `SpinLock` in `monix.execution` btw, but got rid of it because I could not prove that it is better than plain `synchronize`, as the JVM does some weird shit with it.

Reply:
I needed `tryLock()` function because we only need one thread at a time to poll queue so if it is currently occupied the other thread doesn't have to wait. I don't think it's possible with `synchronized` but there might be alternative way to do it (assuming my approach is correct, I don't have full trust in tests with so much concurrency involved)

Reply:
Ah, OK, `tryLock` is an interesting use case.

Reply:
OK, I understood what you did there 🙂 it's very interesting.